### PR TITLE
feat: add plugin to detect native modules

### PR DESCRIPTION
**Pull request by eduardoboucas**
**- Summary**

This PR adds an esbuild plugin to detect modules with native bindings and automatically mark them as external.

**- Test plan**

Added new tests.

**- Description for the changelog**

feat: add plugin to detect native modules

**- A picture of a cute animal (not mandatory but encouraged)**

unnamed

### CODE REVIEW

File: src/runtimes/node/native_modules/detector.js
Comment (eduardoboucas) [inline src/runtimes/node/native_modules/detector.js:2]: I've opted for inlining this check over using [`is-native-module`](https://www.npmjs.com/package/is-native-module) because we also want to look for certain packages in `devDependencies`.

File: src/runtimes/node/native_modules/plugin.js
Comment (erezrokah) [inline src/runtimes/node/native_modules/plugin.js:27]: [sand] can we extract the regex to a const with a meaningful name so it is clear what we're filtering?

File: src/runtimes/node/native_modules/plugin.js
Comment (erezrokah) [inline src/runtimes/node/native_modules/plugin.js:36]: [sand] Should we put the stop condition (`parentDirectory !== directory`) here?

**Review by erezrokah — CHANGES_REQUESTED** [review]
@eduardoboucas nice work.

Some comments, nothing major

File: src/runtimes/node/native_modules/plugin.js
Comment (ehmicky) [inline src/runtimes/node/native_modules/plugin.js:27]: Would merging the `packageName` regExp to `filter` improve performance?

File: src/runtimes/node/native_modules/detector.js
Comment (eduardoboucas) [reply src/runtimes/node/native_modules/detector.js:2]: > Can we merge dependencies and devDependencies and pass the merged object to is-native-module?

Since this is such a hot function, I'd prefer to avoid the cost of creating a new object each time, if we can help it. The impact should be really minimal, but we're trying to heavily optimise this function.

> We could also submit a PR to is-native-module to check for dev dependencies too?

Sounds good! I'll do that.

File: src/runtimes/node/native_modules/plugin.js
Comment (eduardoboucas) [reply src/runtimes/node/native_modules/plugin.js:36]: That means having `parentDirectory` outside the scope of the loop and initialising them both together, which feels a bit odd? What do you think?

File: src/runtimes/node/native_modules/plugin.js
Comment (eduardoboucas) [reply src/runtimes/node/native_modules/plugin.js:27]: Absolutely!

Comment (eduardoboucas) [reply src/runtimes/node/native_modules/plugin.js]: From my tests, no. I think this is because the regex is more complex, which gets more expensive since it'll be executed for each import. Also, the regex we're currently using for filtering already filters out relative and absolute paths, so the number of callback invokations that we'd reduce by using `packageName` would be minimal.

Comment (eduardoboucas) [reply src/runtimes/node/native_modules/plugin.js]: Done in https://github.com/netlify/zip-it-and-ship-it/pull/418/commits/f47f22206c4e80d9e921b7b1e2af96d8dc95d379.

Comment (eduardoboucas) [reply src/runtimes/node/native_modules/plugin.js]: And we'd still need to run `packageName.exec` inside the callback, because we need to extract the directory from the package name.

File: src/runtimes/node/native_modules/plugin.js
Comment (erezrokah) [reply src/runtimes/node/native_modules/plugin.js:40]: Good point. Since `directory` is already in the outer scope we could be consistent with `parentDirectory` too.
I'm good either way.

File: src/runtimes/node/native_modules/detector.js
Comment (erezrokah) [reply src/runtimes/node/native_modules/detector.js:2]: I'm good with inlining to unblock and using the new `is-native-module` once the PR lands

File: src/runtimes/node/native_modules/plugin.js
Comment (ehmicky) [reply src/runtimes/node/native_modules/plugin.js]: :+1:

**Review by ehmicky — APPROVED** [review]

File: src/runtimes/node/native_modules/plugin.js
Comment (eduardoboucas) [reply src/runtimes/node/native_modules/plugin.js:40]: It's just that placing `parentDirectory` in the outer scope means duplicating the `path.dirname` assignment: once outside, to satisfy the loop condition, and a second time inside to calculate the parent directory of each iteration, unless I'm missing what you're suggesting.

Comment (erezrokah) [reply src/runtimes/node/native_modules/plugin.js:40]: Yes, you're right. We can keep it as is 👍

**Review by erezrokah — APPROVED** [review]

**Pull request merged**
